### PR TITLE
feat: access Advanced Search page from New Search UI

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [ ] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the support for accessing Advanced Search from New Search UI. The major change is in `IndexPage` where we configure the route for `SearchPage`.  The result is URLs matching `/page/search`, `searching.do` or `/page/advancedsearch` will link to New Search UI.

The props of `SearchPage` was also updated to accept the ID of an Advanced Search which will used to control the mode of `SearchPage` in future work.

https://user-images.githubusercontent.com/47203811/131597921-26603e7a-1468-4c68-8529-df69aa9afd5c.mp4

### CODE REVIEW

File: react-front-end/tsrc/mainui/IndexPage.tsx [hunk 1]
```diff
@@ -204,6 +167,27 @@ export default function IndexPage() {
       );
     };
 
+    // If the URL is the new Advanced Search path, get ID from the path.
+    // If it's the old one, get ID from query param.
+    const getAdvancedSearchIdFromLocation = (
```
Comment (PenghaiZhang) [inline react-front-end/tsrc/mainui/IndexPage.tsx:172]: Do we want to move this function to `AdvancedSearchModule` ?

File: react-front-end/tsrc/mainui/IndexPage.tsx [hunk 2]
```diff
@@ -204,6 +167,27 @@ export default function IndexPage() {
...
+      location: Location
+    ): string | undefined => {
+      if (location.pathname.match(newAdvancedSearchPagePath)) {
+        const advancedSearchPagePath = /(.+)(\/page\/advancedsearch\/)(.+)/;
+        // The third group is the UUID, and use `reverse` to destructure the array.
+        const [advancedSearchId] =
+          window.location.pathname.match(advancedSearchPagePath)?.reverse() ??
+          [];
```
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/IndexPage.tsx:180]: How about just:

```suggestion
        const advancedSearchId =
          window.location.pathname.match(advancedSearchPagePath)?.pop();
```

File: react-front-end/tsrc/mainui/IndexPage.tsx [hunk 3]
```diff
@@ -204,6 +167,27 @@ export default function IndexPage() {
...
+
+        return advancedSearchId;
+      }
+
+      const currentParams = new URLSearchParams(location.search);
+      const legacyAdvancedSearchId = currentParams.get("in");
+      return legacyAdvancedSearchId?.startsWith("P")
+        ? legacyAdvancedSearchId.substring(1)
+        : undefined;
+    };
```
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/IndexPage.tsx:190]:
```suggestion
    };
    
```

File: react-front-end/tsrc/mainui/IndexPage.tsx (hunk 2 above)
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/IndexPage.tsx:180]: How about just:

```suggestion
        const advancedSearchId =
          window.location.pathname.match(advancedSearchPagePath)?.pop();
```

File: react-front-end/tsrc/mainui/IndexPage.tsx (hunk 1 above)
Comment (edalex-ian) [reply react-front-end/tsrc/mainui/IndexPage.tsx:172]: Sounds like a good idea. :+1:

File: react-front-end/tsrc/mainui/IndexPage.tsx [hunk 4]
```diff
@@ -216,13 +200,26 @@ export default function IndexPage() {
...
+              return (
+                <SearchPage
+                  {...mkRouteProps(p)}
+                  advancedSearchId={getAdvancedSearchIdFromLocation(
+                    window.location
+                  )}
+                />
+              );
             }
-
```
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/IndexPage.tsx:225]: Be nice to keep this blank line. But :man_shrugging:

**Review by edalex-ian — APPROVED** [review]
Looks good @PenghaiZhang , just some minor bits. But merge when ready / when we can get CodeBuild happy again.

File: react-front-end/tsrc/mainui/routes.tsx [hunk 5]
```diff
@@ -164,3 +164,9 @@ export const routes: Routes = {
     to: (uuid: string, version: number) => `/items/${uuid}/${version}/`,
   },
 };
+
+// The component of Search UI varies, depending on the path and whether New Search UI is enabled.
+// So only export their paths.
+export const OLD_SEARCH_PATH_PATH = "/searching.do";
+export const NEW_SEARCH_PATH_PATH = "/page/search";
+export const NEW_ADVANCED_SEARCH_PATH_PATH = "/page/advancedsearch/";
```
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/routes.tsx:172]: What's with all the suffixes of `_PATH_PATH`? Why not just `_PATH`?

File: react-front-end/tsrc/mainui/IndexPage.tsx [hunk 6]
```diff
@@ -216,11 +179,31 @@ export default function IndexPage() {
...
+            const newSearchEnabled: boolean =
+              typeof renderData !== "undefined" && renderData?.newSearch;
+            const location = window.location;
+
+            // If the path matches new Search page path, use `SearchPage`.
+            // If the path matches old Search page path or new Advanced search path, and new
+            // Search UI is enabled, use `SearchPage`.
+            // Use `LegacyContent` for others.
+            if (
+              location.pathname.match(NEW_SEARCH_PATH_PATH) ||
```
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/IndexPage.tsx:197]: Should this also include `NEW_ADVANCED_SEARCH_PATH_PATH` now?

File: react-front-end/tsrc/mainui/routes.tsx [hunk 7]
```diff
@@ -103,9 +104,12 @@ export const legacyPageUrl = (to?: string | ToFunc | ToVersionFunc): string => {
 };
 
 export const routes: Routes = {
-  AdvancedSearch: {
+  OldAdvancedSearch: {
     to: (uuid: string) => `/advanced/searching.do?in=P${uuid}&editquery=true`,
   },
+  NewAdvancedSearch: {
+    to: (uuid: string) => `/page/advancedsearch/${uuid}`,
```
Comment (edalex-ian) [inline react-front-end/tsrc/mainui/routes.tsx:111]: Should probably use the new `NEW_ADVANCED_SEARCH_PATH_PATH` constant up here for the start of this string..

**Review by edalex-ian — APPROVED** [review]
Looks good @PenghaiZhang . Let's merge this in. (I'm keen to merge this into my branch to build on. :wink: )

**Comment by edalex-ian** [comment]
And no need to worry about the flaked `MyResourcesTest`. Just merge on. :wink:

**Pull request merged**